### PR TITLE
Return literal string from sprintf on literal input

### PR DESCRIPTION
**Pull request by Khartir**
I noticed this when updating `phpstan-doctrine`, which now requires `literal-string` as arguments to `where`. We have a couple of where-clauses with dynamic columns, which we build using `sprintf`. But `sprintf` does not return a `literal-string`, even if all inputs are `literal-string`.

### CODE REVIEW

**Comment by craigfrancis** [comment]
Hi @Khartir, can you share some examples of your dynamic columns?

The output from `sprintf()` can get a bit complicated, and I'd rather we follow the [patch](https://github.com/craigfrancis/php-is-literal-rfc/blob/main/patches/8.1.2-main.diff) for the [is_literal RFC](https://wiki.php.net/rfc/is_literal), which does not support `sprintf()` (the RFC will be re-considered after we have some feedback from static analysis tools like PHPStan).

Normally I would have an allow-list of safe `literal-string` values, so attackers cannot simply sort by any field; e.g.

    $fields = ['name', 'address', 'email'];
    
    $field_id = array_search($unsafe_field, $fields);
    
    $field = $fields[$field_id];

Or if you want to use aliases for those fields:

    $fields = [
      'name'    => 'u.full_name',
      'email'   => 'u.email_address',
      'address' => 'u.postal_address',
    ];
    
    $field = ($fields[$unsafe_field] ?? 'u.full_name');

Both of these return a `literal-string` that can be used like this:

    $qb
      ->select('u')
      ->from('User', 'u')
      ->where($field . ' = :user_value')
      ->setParameter('user_value', $user_value);

File: tests/PHPStan/Type/Php/data/sprintf.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace PHPStan\Type\Php\Test;
+
+use function PHPStan\Testing\assertType;
+
+$foo = 'foo';
+assertType("'foo'", sprintf('%s', $foo));
+
+function foo (string $nonLiteral): void {
+	assertType('non-empty-string', sprintf('%s', $nonLiteral));
```
Comment (staabm) [inline tests/PHPStan/Type/Php/data/sprintf.php:11]: When the format is just `%s` and we dont know whether $nonLiteral was a non-empty-string before, how can it be a non-empty-string after the sprintf-call?

File: tests/PHPStan/Type/Php/data/sprintf.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+
+/** @var literal-string $literal */
+$literal = '';
+assertType('literal-string&non-empty-string', sprintf('%s', $literal));
+
+function bar (int $int): void {
+	assertType('literal-string&non-empty-string', sprintf('%d', $int));
```
Comment (staabm) [inline tests/PHPStan/Type/Php/data/sprintf.php:19]: Imo it can only return a literal string when the args to sprintf were literal.. but a regular int is not literal per se

Comment (craigfrancis) [reply tests/PHPStan/Type/Php/data/sprintf.php:19]: Thanks Markus; and for anyone wondering why (frustrating as it is), we discussed this in [Issue 6033](https://github.com/phpstan/phpstan/issues/6033).

**Comment by Khartir** [comment]
> Hi @Khartir, can you share some examples of your dynamic columns?

Sure. Basically our code looks like this:
`->where(sprintf('ac.%s = :id', static::$idField))`
`static::$idField` is basically a `const` that does not contain user input and we use `sprintf` instead of string concatination for coding standard reasons and better readability.

> The output from `sprintf()` can get a bit complicated, and I'd rather we follow the [patch](https://github.com/craigfrancis/php-is-literal-rfc/blob/main/patches/8.1.2-main.diff) for the [is_literal RFC](https://wiki.php.net/rfc/is_literal), which does not support `sprintf()` (the RFC will be re-considered after we have some feedback from static analysis tools like PHPStan).

If that's the case, the only step is of course to close this PR. But I I'm curious under which circumstances the output of `sprintf()`  should not be considered a literal string, if all inputs are literal strings.

File: tests/PHPStan/Type/Php/data/sprintf.php (hunk 2 above)
Comment (Khartir) [reply tests/PHPStan/Type/Php/data/sprintf.php:19]: Yeah I think I got a bit carried away. I'll fix it, unless the rest of the discussion leads to this PR being closed.

File: tests/PHPStan/Type/Php/data/sprintf.php (hunk 1 above)
Comment (Khartir) [reply tests/PHPStan/Type/Php/data/sprintf.php:11]: This is basically a test I added before I added my changes to make sure I wouldn't break anything. As far as I can tell, there were no tests for the return type of `sprintf()` before.
The return type of `non-empty-string` confused me to, but the current logic is incredibly simple there and this edge case of "the format is just placholder(s)" is probably not worth the effort of checking this.
I was thinking of writing a comment in the test, but now I think it's probably simpler to just add a string to the format and use `'foo %s'` or something along those lines.

**Comment by craigfrancis** [comment]
When you say coding standard reasons, can you explain a bit more?

As to readability, I wonder if it's because I'm more used to it, but I find the string concatenation easier to read, with it not needing to call a function, or needing to look to the right to see what would replace "%s":

```diff
- ->where(sprintf('ac.%s = :id', static::$idField))
+ ->where('ac.' . static::$idField . ' = :id')
```

---

With the native PHP implementation, Joe Watkins tried updating `sprintf()` to do what you suggest, and I can't remember all of the points (e.g. possible performance impact), but we decided that it's output can no longer be called a programmer-defined-string (aka a literal), because the formatting can change the input values.

It's why we limited ourselves to the functions `str_repeat()`, `str_pad()`, `implode()` and `join()`, which we could argue were no more than glorified concatenation functions (that's not to say it's a done deal, if a good argument can be made, then we would re-consider).

As to the formatting modifications, while they would *probably* be fine, it's hard to be 100% confident it will be perfectly safe in every single context, like HTML/SQL/CLI/etc... the example I came up with, while hard to think of an actual exploit, is how the locale can change the decimal place from a "." (fine for SQL) to a "," (probably a parse error for SQL):

```php
setlocale(LC_ALL, 'de_DE.UTF-8');
var_dump(sprintf('%.3f', '1.23')); // Output = "1,230"
```

**Comment by Khartir** [comment]
> When you say coding standard reasons, can you explain a bit more?

We use an extended, modified version of the [Symfony coding standard](https://symfony.com/doc/current/contributing/code/standards.html#structure). They define:
> Exception and error message strings must be concatenated using [sprintf](https://secure.php.net/manual/en/function.sprintf.php);

We basically extended this to all method calls (I think, I'm not actually sure).

Regarding readability, I personally find it more convenient to be able to easily read `'ac.%s = :id'` as "I want to compare equality of some field with some value" and then checking what the field actually is. But yes a lot of that comes down to getting used to one style over the other.

I have to admit I'm very ambivalent about all of this. I see your point regarding the problems of `sprintf` but coming from a project that makes heavy use of it for concatenation I also can't stop thinking the two should be interchangeable.

I guess the best way forward would be to open an issue in https://github.com/phpstan/phpstan and see how many other people would like to use `sprintf` with `literal-string`. What do you think?

**Comment by craigfrancis** [comment]
Thanks @Khartir.

Because this would also affect [Psalm](https://github.com/vimeo/psalm/releases/tag/4.8.0), hopefully [Phan](https://github.com/phan/phan/issues/4618), and maybe PHP itself... I've [created an Issue in my repo for the original RFC](https://github.com/craigfrancis/php-is-literal-rfc/issues/5) (feel free to discuss elsewhere as well).

---

I've done a bit more reading, and I found [this article and the comments interesting](https://tomasvotruba.com/blog/2018/10/11/hi-my-name-is-tom-conctat-vs-in-sprintf-vs-in-string-variables/)... overall it *seems* to be a personal preference thing, which implies a few people might want this.

I should note, at least 2 PHP Internals developers (Dan Ackroyd and George P. Banyard) voted against the original RFC because they did not like it supporting basic string concatenation, they wanted everyone to use new functions `literal_concat()` and `literal_implode()`, as doing so might help with debugging (would throw an exception as soon as a mistake is made). I think they are in the minority, where myself and others believe string concatenation would help adoption (because a lot of existing code does this), and not everyone wants to use a function that throws exceptions (also, I did provide Dan/George [userland implementations of these functions](https://wiki.php.net/rfc/is_literal#support_functions), and they could use tools like [phpcs](https://github.com/squizlabs/PHP_CodeSniffer) to enforce their own stricter coding standards).

Back to `sprint()`, I think I would prefer it didn't return a literal-string, but that's only because I want to keep the implementation as simple as possible, and to only add features if they help and do not cause issues.

**Comment by ondrejmirtes** [comment]
Agreed!
